### ISSUE TITLE
handling missing/non-finite likelihoods

### ISSUE THREAD

**Issue by davharris** [weecology/sad-comparison#187]
I'd like to take a look at this issue before resubmitting.  It may be possible to handle these better than just omitting them from the analysis.

**Comment by ethanwhite** [comment]
Agreed. I'm actually wondering whether this is going to have been fixed by improving the solvers and eliminating the Gentry sets with the bad data that was causing at least some of the failures.

@embaldridge - do you remember roughly how many likelihoods were giving infinite/NA values?

**Comment by davharris** [comment]
I wonder whether it would be better to leave the "bad" values in, rather than making them all blank. That way, we could see whether a given value was "bad" because it was non-finite, missing, non-converged, etc.

**Comment by ethanwhite** [comment]
I'm rerunning both the current `master` and a version that has removed these checks now to see if the improvements have removed these cases.

If some still exist then I agree that we should keep them. That will require either changing the logic of how the different `*_list` objects get handled, or saving the "bad" sites in a separate output object.

**Comment by embaldridge** [comment]
I don't think it was that many, but there may be a closed issue related to
that, or I put a note somewhere on the repository or something like that.
I think it was mostly with the Zipfs and the negative binomials?

Elita Baldridge, Ph.D.

GitHub: embaldridge
Twitter: @elitabaldridge
Website: http://elitabaldridge.org/

**Comment by davharris** [comment]
Negative binomial might be fixed entirely with the new code.  I'll see if we can do something similar for Zipf.

**Comment by embaldridge** [comment]
The Poisson lognormal had some -inf responses: Issue #23 .  Mostly FIA datasets.

**Comment by ethanwhite** [comment]
Looks like that issue might have been fixed in d367a40044d83f11e7d110346aab21ce24a9cdf9.

**Comment by ethanwhite** [comment]
On a related note, we do have a fair number of:

```
Warning: Maximum number of function evaluations has been exceeded.
Optimization terminated successfully.
```

warnings that are being returned while still producing parameter estimates and therefore likelihoods, AICs, etc.

**Comment by davharris** [comment]
Any idea which solvers are causing them? 

We can probably address these by increasing the iteration limit or providing better starting values, but it would be good to know which solvers were having trouble. 

Dave

**Comment by davharris** [comment]
[Stack Overflow just reminded me](https://stackoverflow.com/questions/29229810/optimisation-using-scipy) that this warning could also crop up when the optimum is not finite (e.g. trying to find the optimum of `f(x)=x`).

Given that my suggested changes to the solvers result in unbounded optimization problems, it's possible that this is what's going on.

**Comment by rueuntal** [comment]
Dave - Among these the only one that I'm sure would work is logseries (parameter bounded, function intersects with zero only once within the bound). My test run with Gentry yesterday had inf/nan for PLN and negative binomial at some sites, but not for logseries or Zipf.

**Comment by davharris** [comment]
Okay, good to know. Thanks.

**Comment by ethanwhite** [comment]
`inf` and `nan` records:

* [nan_data.txt](https://github.com/weecology/sad-comparison/files/143114/nan_data.txt)
* [inf_data.txt](https://github.com/weecology/sad-comparison/files/143115/inf_data.txt)

**Comment by ethanwhite** [comment]
I don't think the commit I referenced earlier actually did anything (it just looked up the floating point limit), and it also doesn't make sense to me that these log-likelihoods should be below machine precision. I actually ran this in a sort of funny way, so let me rerun and make sure this is all correct.

**Comment by davharris** [comment]
Okay, let's think about this.  Assuming no bugs in `pln_ll`, I think that the only time we should get `NaN` would be if sigma went negative.  Seems like we could prevent that either with a constrained optimizer (like `scipy.optimize.fmin_l_bfgs_b`) or by working with log(sigma).  My preference is probably for the latter option.

The `-Inf` values for the negative binomial are harder for me to wrap my head around.  Wouldn't the optimizer avoid these values? If it just breaks when it gets non-finite values, we could hack around it by replacing infinities with `1.797693e+308` or whatever Python's upper bound is for floats.  If something else is going on, that would be good to know.

**Comment by ethanwhite** [comment]
@davharris - when you compared the output with that from R what was going on with these values in R?

**Comment by davharris** [comment]
I don't think I've run them in R. I'll see what I can do. 

Dave

**Comment by davharris** [comment]
Looks like I did run the negative binomial values in R: no `-Inf`s in the BBS data.  I must have broken something, because none of those values were -Inf before (in Python *or* in R)

**Comment by ethanwhite** [comment]
Following https://github.com/weecology/macroecotools/pull/37 here's the most recent data with `nan` and `inf` likelihoods:
[naninf_data.txt](https://github.com/weecology/sad-comparison/files/145420/naninf_data.txt)

**Comment by davharris** [comment]
Current guess is that nbinom is breaking when `n` is less than machine precision (1E-16 or whatever).

Example:

```
ab = [2L, 2L, 62L, 8L, 5L, 25L, 7L, 2L, 1L, 94L, 4L, 3L, 106L, 34L, 3L, 2L, 1L, 3L, 20L, 5L, 2L, 9L, 131L, 11L, 1L, 1L, 1L, 133L, 214L, 20L, 2L, 5L]

n, p = md.nbinom_lower_trunc_solver(ab) # n is about 1E-20
md.nbinom_lower_trunc_ll(ab, n, p)  # Returns inf
```

No idea why `n` gets that small, though.  When I use my R code to fit it, I get `n` closer to 1E-5. Maybe it's the starting values?

**Comment by ethanwhite** [comment]
Following https://github.com/weecology/macroecotools/pull/37 here's the most recent data with `nan` and `inf` likelihoods:
[naninf_data.txt](https://github.com/weecology/sad-comparison/files/145420/naninf_data.txt)

**Comment by ethanwhite** [comment]
All of these PLN `nan`s are almost certainly new as well.

**Comment by rueuntal** [comment]
@davharris what's the R function that you use to fit negative binomial? `fitdistr()` seems to be using the same starting values (mean for m, mean^2/(var - mean) for k, which is equivalent to our configuration) and the same optimization method (BFGS) but is more stable than our function for at least some ab. I haven't quite figured out what's going on.

**Comment by rueuntal** [comment]
I think one solution for these non-converging problems could be to switch between different optimization methods and see which one works. The very naive message that I took from my optimization class is that BFGS is one of the most stable methods in many cases, which is why it's implemented in our code. But maybe alternative methods can help us when BFGS fails:
```
from __future__ import division
from macroeco_distributions import *
ab = np.array([  2,   2,  62,   8,   5,  25,   7,   2,   1,  94,   4,   3, 106,
        34,   3,   2,   1,   3,  20,   5,   2,   9, 131,  11,   1,   1,
         1, 133, 214,  20,   2,   5]) # BBS 54130
mu = np.mean(ab)
var = np.var(ab, ddof = 1)
p0 = 1 - mu / var
if p0 < 0: p0 = 10**-5
elif p0 > 1: p0 = 1 - 10**-5
n0 = mu * (1 - p0) / p0
def negbin_func(x):
    return -nbinom_lower_trunc_ll(ab, x[0], x[1])
ans = optimize.fmin_bfgs(negbin_func, x0 = [n0, p0], full_output = 1)
warnflags = ans[6]
print warnflags # Warning for convergence issues
if warnflags is 0:
    n, p = ans[0]
else:
    n, p = optimize.minimize(negbin_func, [n0, p0], method = 'Nelder-Mead').x
print nbinom_lower_trunc_ll(ab, n, p)
```
The log-likelihood that I got using the above code is actually higher than that from `fitdistr()` in R (-121.2 vs -132.2). But right now this approach has (at least) two problems: 1. there's no guarantee that Nelder-Mead would always converge in cases where BFGS doesn't. If there are still inf/nan, we'd probably have to switch for yet another alternative. 2. I had to revert the log/logit transformation for n and p because they caused problems for Nelder-Mead, but this may not be desirable for other sites. 

Comments?

**Comment by rueuntal** [comment]
Following my previous comment, to solve problem 1 we can probably write a function that tries BFGS, Nelder-Mead, Powell and CG (the four methods in optimize.minimize that are unconstrained and don't require additional inputs) subsequently, moving to the next method only when the previous one(s) fail(s). If all four fail, the function can then admit failure and return nan. What do you folks think?

**Comment by ethanwhite** [comment]
> Following my previous comment, to solve problem 1 we can probably write a function that tries BFGS, Nelder-Mead, Powell and CG (the four methods in optimize.minimize that are unconstrained and don't require additional inputs) subsequently, moving to the next method only when the previous one(s) fail(s). If all four fail, the function can then admit failure and return nan. What do you folks think?

At first glance this sounds good to me, but... I am still confused as to why so many `nan`s started showing up. Looking back at @embaldridge's results from late 2015, we don't have anything like the number of `nan`s that we're seeing from my last few runs, so I'm guessing we broke something somewhere and should figure that out before spending a bunch of time tweaking the optimizers. The first BBS route I show `nan`ing for the PLN is 11006, then 11057 and 11068. Can you check and see if they are `nan`ing for you to see if it's something I did locally or something in the code?

**Comment by rueuntal** [comment]
I've submitted a pull request for `macroeco_distributions` that incorporates @davharris ' suggestion on constraining sigma to be > 0 in PLN. That seems to have fixed the problem in PLN in the few cases that I tested for BBS. I just did a rerun to make sure. The modified `pln_solver()` returns reasonable estimates for mu and sigma for the three sites you mentioned, and `pln_ll()` finite likelihoods. 

My previous (3?) comments are for negative binomial only. Sorry about the confusion!

**Comment by davharris** [comment]
I'm at the airport and can't do much today to help today, but I agree with @ethanwhite that this is likely a recent problem.  I'm pretty sure there weren't nearly as many failures for negative binomial as of a947e61f66d6248be1e8889583bcf15147d941b9.  I'd rather fix the root cause, if possible, than try to work around it by trying different optimizers.

I think that the results I posted [above](https://github.com/weecology/sad-comparison/issues/187#issuecomment-188468756) suggest that it's a problem with the objective function, not the optimizer.

**Comment by rueuntal** [comment]
Hmmm for PLN I think I've found where the problem lies, though don't exactly understand why. The problem is the `_argcheck` method that I added, which checks if the parameters are reasonable (in this case, sigma > 0) and if the x value is within bound (>=1 if lower-truncated, >=0 if not). If x is out of bound, it returns a 0. If sigma is negative, it turns nan. 

Replacing the current code under `_argcheck` with `return 1` (ie always pass) gets us back to reasonable estimates in `pln_solver()`. What's baffling me is that `pln_solver()` can return negative estimates for sigma when the checks ARE PRESENT, but reasonable values when they are not.

**Comment by davharris** [comment]
What happens if argcheck expects three arguments in args and only gets two, like in untrunc_pmf(x_i, mu, sigma)?

Dave

**Comment by davharris** [comment]
Best guess is that self.a isnt being set correctly , causing zeros to have zero probability (negative infinity likelihood)

Dave

**Comment by rueuntal** [comment]
Having zero probability when x is out of bound is the desired behavior and
identical to that of built-in distributions.

And there are no zeros in the spab files, right?

**Comment by rueuntal** [comment]
Do you mean, say, pln.pmf(1, 1, 2)? Python would throw an error message.

**Comment by rueuntal** [comment]
OK, so the check for sigma is the one causing the problem. Removing this line restores the old behavior of `pln_solver()`, though I'd be hesitant to go this route because it feels appropriate to have checks for inputs. How do you feel about this?

And the problem with the negbin solver is probably something else, since `_argcheck` has been there for negbin for a while. I'll look deeper into it.

**Comment by rueuntal** [comment]
For negbin, I wonder if transformation on n and p is partly the problem. For the three BBS sites with inf log-likelihood: 
```
from __future__ import division
import macroeco_distributions as md
import numpy as np
from scipy import optimize

def negbin_solver_untransformed(ab):
    """Given abundance data, solve for MLE of negative binomial parameters n and p"""
    mu = np.mean(ab)
    var = np.var(ab, ddof = 1)
    p0 = 1 - mu / var
    n0 = mu * (1 - p0) / p0
    def negbin_func(x): 
        return -md.nbinom_lower_trunc_ll(ab, x[0], x[1])
    n, p = optimize.fmin(negbin_func, x0 = [n0, p0])
    return n, p

def import_abundance(datafile):
    """Imports raw species abundance .csv files in the form: Site, Year, Species, Abundance."""
    raw_data = np.genfromtxt(datafile, dtype = "S15,i8,S50,i8", names = ['site','year','sp','ab'], delimiter = ",",comments = "#")
    return raw_data

def print_negbin_results(data, site):
    ab = data[data['site'] == site]['ab']
    n1, p1 = md.nbinom_lower_trunc_solver(ab)
    print "Par estimation with transformation: ", n1, p1
    print "Loglikelihood with transformation: ", md.nbinom_lower_trunc_ll(ab, n1, p1)
    
    n2, p2 = negbin_solver_untransformed(ab)
    print "Par estimation without transformation: ", n2, p2
    print "Loglikelihood without transformation: ", md.nbinom_lower_trunc_ll(ab, n2, p2)
    
data_dir = './sad-data/bbs_spab.csv'
raw_data = import_abundance(data_dir)

print_negbin_results(raw_data, '54130')
print_negbin_results(raw_data, '92061')
print_negbin_results(raw_data, '6101')
```

**Comment by davharris** [comment]
Yup. Still not sure why `n` goes below 1E-16 with this Python code when the (mostly-equivalent) R code doesn't get anywhere near as small, though. I need to do some more tests.

**Comment by davharris** [comment]
>OK, so the check for sigma is the one causing the problem. Removing this line restores the old behavior of pln_solver(), though I'd be hesitant to go this route because it feels appropriate to have checks for inputs. How do you feel about this?

I'd rather identify and resolve the underlying problem, if possible.

Is it possible that there's a problem with the indexing? Scipy (wisely, in my opinion) names all of the arguments to _argcheck ([example](https://github.com/scipy/scipy/blob/master/scipy/stats/_discrete_distns.py#L160)) so that it's easier to keep track of which variable gets which bounds.

Also, this could just be because I don't speak much Python, but modifying `self.a` inside _argcheck creeps me out. It's probably safe in this case, depending on how often _argcheck gets called, but I'm pretty sure it means that the global value of `pln` changes depending on whether the last call was truncated or not.

**Comment by davharris** [comment]
I think I figured out what's going on.  Should have read the warning messages I got before more carefully.

Inputs:

```
ab = [2L, 2L, 62L, 8L, 5L, 25L, 7L, 2L, 1L, 94L, 4L, 3L, 106L, 34L, 3L, 2L, 1L, 3L, 20L, 5L, 2L, 9L, 131L, 11L, 1L, 1L, 1L, 133L, 214L, 20L, 2L, 5L]
n, p = md.nbinom_lower_trunc_solver(ab)
```

Output:

```
/Users/davidharris/anaconda/lib/python2.7/site-packages/scipy/stats/_distn_infrastructure.py:815: RuntimeWarning: divide by zero encountered in log
  return log(self._sf(x, *args))
/Users/davidharris/anaconda/lib/python2.7/site-packages/scipy/optimize/optimize.py:462: RuntimeWarning: invalid value encountered in subtract
  numpy.max(numpy.abs(fsim[0] - fsim[1:])) <= ftol):
Warning: Maximum number of function evaluations has been exceeded.
```

Here's what's going on:

1) when we use the log/logit transforms, the solver manages to get to the extreme areas of parameter space, like `n=1E-20`.
2) When that happens, essentially all of the probability mass is at zero, and none is left for the untruncated part of the distribution.
3) The optimizer gets stuck in this region somehow, and runs out of iterations before it can escape

I think we just need a different optimizer. Or we could tell the optimizer never to let `n` go below some particularly extreme values like 1E-16.

**Comment by rueuntal** [comment]
>Is it possible that there's a problem with the indexing? Scipy (wisely, in my opinion) names all of the arguments to _argcheck (example) so that it's easier to keep track of which variable gets which bounds.

Brilliant, Dave! I still have no idea why but that seems to work. I'll go ahead and update `_argcheck` in other distributions too.

>Also, this could just be because I don't speak much Python, but modifying self.a inside _argcheck creeps me out. It's probably safe in this case, depending on how often _argcheck gets called, but I'm pretty sure it means that the global value of pln changes depending on whether the last call was truncated or not.

I may be confused about what you meant, but I think `pln` gets reset every time it is called, because `lower_trunc` always has to be defined. For example if we run 
`pln.pmf(1, 1, 2, lower_trunc = True)
pln.pmf(0, 1, 2, lower_trunc = True)
pln.pmf(0, 1, 2, lower_trunc = False)
pln.pmf(1, 1, 2, lower_trunc = False)`
it returns 
`0.18951492766270298
0.0
0.25535406091980023
0.1411215212791096`

**Comment by rueuntal** [comment]
>Is it possible that there's a problem with the indexing? Scipy (wisely, in my opinion) names all of the arguments to _argcheck (example) so that it's easier to keep track of which variable gets which bounds.

Sorry about the previous (deleted) comment. I thought it worked but it was because in my local `macroeco_distributions` I have implemented your suggestion on log-transforming sigma. So the problem is probably not indexing, though I'd be happy to get them changed for better style. 

>Also, this could just be because I don't speak much Python, but modifying self.a inside _argcheck creeps me out. It's probably safe in this case, depending on how often _argcheck gets called, but I'm pretty sure it means that the global value of pln changes depending on whether the last call was truncated or not.

Just in case that you didn't see my previous comment before I deleted it - I think `pln` gets reset every time it is called, because `lower_trunc` always has to be specified. For example if you call `pln.pmf(0, 1, 2, lower_trunc = True)`, then `pln.pmf(0, 1, 2, lower_trunc = False)`, then `pln.pmf(0, 1, 2, lower_trunc = True)` again, it will return 0, some value, 0. Or did you mean something else?

**Comment by rueuntal** [comment]
OK I've attempted a fix with the [pull request](https://github.com/weecology/macroecotools/pull/39). It's kind of ad hoc but will probably fix most if not all of the nan/inf problems. 

For negative binomial I'm implementing Dave's suggestion on constraining log(n) to prevent n from getting too close to zero. It's been tested using the three problematic BBS sites (54130, 92061, 6101) and n doesn't even approach the lower bound when it's there. 

I'm still perplexed by what's going on with `_argcheck` in PLN. But for the time being, the problem can be fixed by enforcing sigma > 0 in the solver. I'm implementing Dave's suggestion on log transformation, as well as a lower limit identical to the negbin solver to reduce the numbers of runtime warnings when the solver approaches zero. 

@davharris @ethanwhite could you guys take a look when you've got the time?

**Comment by ethanwhite** [comment]
I love working with really smart folks committed to getting it right. You two are awesome.

The improvements in @rueuntal's PR have put us back in good shape:

```
ethan@stice:~/Dropbox/Research/SADComparison/sad-data/chapter1 (refactor)$ grep 'nan\|inf' *_likelihoods.csv
cbc_likelihoods.csv:L15187,61,6934,-340.11174061952238,-332.32045086740328,inf,-362.47081604523072
fia_likelihoods.csv:12025000038,11,33,-22.841450451929525,-19.760025651198585,nan,-25.657742363750003
fia_likelihoods.csv:131305000180,17,34,-23.824308614661692,-23.668528863381479,nan,-25.522100388768173
fia_likelihoods.csv:135311000006,10,20,-14.058101095604163,-13.92471182358352,nan,-15.104087441684038
fia_likelihoods.csv:14063000049,10,25,-17.715026765859729,-16.521361509885814,nan,-19.592367405604783
fia_likelihoods.csv:15073000111,10,21,-14.78296485058199,-14.614299779244886,nan,-15.935519910956794
fia_likelihoods.csv:15111000048,11,23,-16.088070319334623,-16.064296143434891,nan,-17.239127602127997
fia_likelihoods.csv:15133000110,10,25,-17.791987806995859,-16.420820850756922,nan,-19.733956372208354
fia_likelihoods.csv:16059000040,10,21,-14.78296485058199,-14.614299779245211,nan,-15.935519910956796
fia_likelihoods.csv:171003020045,10,20,-13.834957544289953,-14.023361241409406,nan,-14.637323052834077
fia_likelihoods.csv:171077020210,10,21,-14.78296485058199,-14.614299779245211,nan,-15.935519910956794
fia_likelihoods.csv:172083020072,10,25,-17.491883214545521,-16.698092423545738,nan,-19.179206390863804
fia_likelihoods.csv:181055020064,11,23,-16.088070319334623,-16.064296143434891,nan,-17.239127602127997
fia_likelihoods.csv:182123010230,11,21,-14.094682037149116,-14.204330899981791,inf,-14.431070784238516
fia_likelihoods.csv:192057020033,11,33,-22.451733700789497,-20.179945199611989,nan,-24.988805452414063
fia_likelihoods.csv:202107020350,10,25,-17.791987806995859,-16.420820850755426,nan,-19.733956372208358
fia_likelihoods.csv:212127000034,11,22,-15.831862560995095,-15.080383086599422,nan,-17.370631439901704
fia_likelihoods.csv:213013000008,11,23,-16.088070319334623,-16.064296143434891,nan,-17.239127602127997
fia_likelihoods.csv:215057000013,10,21,-15.252968479827727,-14.281830224385487,nan,-16.87479822814436
fia_likelihoods.csv:216177000044,11,23,-16.088070319334623,-16.064296143434891,nan,-17.239127602127997
fia_likelihoods.csv:222047000053,10,25,-17.491883214545521,-16.698092423544715,nan,-19.179206390863804
fia_likelihoods.csv:223069000106,10,21,-14.782964850581989,-14.614299779245211,nan,-15.935519910956792
fia_likelihoods.csv:242005000979,10,21,-14.78296485058199,-14.614299779244886,nan,-15.935519910956794
fia_likelihoods.csv:242041001114,11,33,-22.751838293239839,-19.893022971557215,nan,-25.504595083290219
fia_likelihoods.csv:264159020016,10,21,-14.78296485058199,-14.614299779245211,nan,-15.935519910956796
fia_likelihoods.csv:341037000129,11,23,-16.088070319334623,-16.064296143434891,nan,-17.239127602127997
fia_likelihoods.csv:373025000068,14,28,-19.666175028294564,-19.48126794102517,nan,-21.114216730847925
fia_likelihoods.csv:374021000031,17,51,-34.647572942617764,-31.290753303031408,nan,-38.531736272436476
fia_likelihoods.csv:374115000026,13,39,-26.434726755421742,-23.958596419186815,nan,-29.360862289055476
fia_likelihoods.csv:392105000069,11,22,-15.361858931749362,-15.361715249623547,nan,-16.401938847038533
fia_likelihoods.csv:429071000582,10,20,-13.834957544289955,-14.023361241409406,nan,-14.637323052834077
fia_likelihoods.csv:452051000177,10,25,-17.715026765859733,-16.521361509885523,nan,-19.59236740560479
fia_likelihoods.csv:471047000055,10,25,-17.791987806995859,-16.420820850756922,nan,-19.733956372208358
fia_likelihoods.csv:471077000030,10,21,-14.78296485058199,-14.614299779245211,nan,-15.935519910956796
fia_likelihoods.csv:471157000055,15,30,-20.969932864439762,-20.916470272212717,nan,-22.412100594500224
fia_likelihoods.csv:473021000005,10,25,-17.791987806995863,-16.420820850754531,nan,-19.733956372208354
fia_likelihoods.csv:473027000007,14,29,-20.395320233861376,-20.187005928099779,nan,-21.949151214752796
fia_likelihoods.csv:473055000026,10,21,-14.78296485058199,-14.614299779245211,nan,-15.935519910956794
fia_likelihoods.csv:475073000013,15,30,-20.969932864439766,-20.916470272212717,nan,-22.412100594500227
fia_likelihoods.csv:51117000075,10,21,-14.78296485058199,-14.614299779245211,nan,-15.935519910956797
fia_likelihoods.csv:515173000054,11,23,-16.088070319334623,-16.064296143434891,nan,-17.239127602127997
fia_likelihoods.csv:542023001399,11,22,-15.361858931749362,-15.361715249623547,nan,-16.401938847038533
fia_likelihoods.csv:552107020150,11,23,-16.088070319334623,-16.064296143434404,nan,-17.239127602127997
```

There are now only 43 sites with issues with issues, all with the negative binomial, 42 from FIA and one from CBC. All of the FIA cases have N ~ 2*S, which will definitely be on the extreme end of things. I'm not sure what's going on with the CBC route. This is basically what I remember us having before `argcheck` started messing with us.

I don't understand the optimization stuff as well as either of you, so I'm planning on leaving the initial review of https://github.com/weecology/macroecotools/pull/39/ to @davharris unless he'd prefer I handle it (in which case just let me know).

**Comment by rueuntal** [comment]
I think I know what's going on for some of the FIA sites - it's a minor bug in `nbinom_lower_trunc_solver()`. But let me make sure.

**Comment by rueuntal** [comment]
I just committed the change. It now works on my end for the CBC site, and a few of the FIA sites I tested. @ethanwhite could you give it another try?

**Comment by ethanwhite** [comment]
Yep, I'm on it.

**Comment by davharris** [comment]
Nice! I'm not thrilled with the arbitrary thresholds at 1E-16; I'd really rather understand why the optimizer was misbehaving. But I don't think it will cause any problems because the optimal value is apparently never that small. 

I agree with the most recent change to the nbinom solver--nice catch. I'll try to find time to look at the other changes in #39 soon, although it really is a lot of code in an unfamiliar language, so @ethanwhite's feedback would probably be good too. 

Dave

**Comment by rueuntal** [comment]
Thanks guys for the quality control!

**Comment by ethanwhite** [comment]
I'm now down to the two `inf`'s!

```
ethan@stice:~/Dropbox/Research/SADComparison/sad-data/chapter1 (refactor)$ grep 'nan\|inf' *_likelihoods.csv
cbc_likelihoods.csv:L15187,61,6934,-340.11174061952238,-332.32045086740328,inf,-362.47081604523072
fia_likelihoods.csv:182123010230,11,21,-14.094682037149116,-14.204330899981791,inf,-14.431070784238516
```

But you said you had the CBC site working on your end so I'm not sure why I'm seeing something different.

**Comment by rueuntal** [comment]
Hmmm, here's what I got for the CBC site: 
```
site = 'L15187'
ab = raw_data[raw_data['site'] == site]['ab']
len(ab)
61
sum(ab)
6934
n, p = md.nbinom_lower_trunc_solver(ab)
print n, p
3.13740290304e-07 0.00132698262057
print md.nbinom_lower_trunc_ll(ab, n, p)
-340.11172857
```
And for the FIA site:
```
site = '182123010230'
ab = raw_data[raw_data['site'] == site]['ab']
print len(ab), sum(ab)
11 21
n, p = md.nbinom_lower_trunc_solver(ab)
print n, p
1.3095648159e-05 0.307249735995
print md.nbinom_lower_trunc_ll(ab, n, p)
-14.0946917775
```

**Comment by ethanwhite** [comment]
Hmmmm, OK, let me test them directly.

**Comment by ethanwhite** [comment]
Interesting. Looks like we are getting different answers on different computers. I uninstalled and reinstalled after checking the commit log on the PR, so I should have the right version.

```
In [15]: site = 'L15187'
In [16]: ab = raw_data[raw_data['site'] == site]['ab']
In [17]: ab
Out[17]: 
array([  15,   12,    7,  295,    5,   66,   14,   48,   14,   71,   29,
         25,    1,  126,  102,    8,   69,   89,   35,    6,  111,    3,
         13,    5,   21,   22,  145,   13,  115,   13,    7,  818,    1,
        157,   66,    2,   21,   57,   70,   19,   40,   25,  414,  263,
        614,  124,  355,  130,  250,    1, 1364,   74,    4,   40,   92,
         11,   75,   13,    7,  290,   32])
In [18]: n, p = md.nbinom_lower_trunc_solver(ab) 
In [19]: print n, p
5.30486338138e-16 0.909513472349
In [20]:  print md.nbinom_lower_trunc_ll(ab, n, p)
inf
```

and

```
In [21]: site = '182123010230'

In [22]: raw_data = import_abundance('sad-data/chapter1/fia_spab.csv')

In [23]: ab = raw_data[raw_data['site'] == site]['ab']

In [24]: ab
Out[24]: array([6, 2, 1, 1, 1, 4, 1, 1, 2, 1, 1])

In [25]: len(ab), sum(ab)
Out[25]: (11, 21)

In [26]: n, p = md.nbinom_lower_trunc_solver(ab)

In [27]: n, p
Out[27]: (2.0292143741963476e-16, 0.30466039212444623)

In [28]: md.nbinom_lower_trunc_ll(ab, n, p)
Out[28]: inf
```

**Comment by rueuntal** [comment]
That's a bit... disconcerting. Could it be different versions of `scipy` (and hence the optimizer)? Mine is 0.15.1.

**Comment by ethanwhite** [comment]
Yeah... never what you want to see. We do have a version difference for scipy:

```
In [37]: scipy.__version__
Out[37]: '0.16.1'
```

It could also be a machine precision difference I guess:

```
In [39]: import sys

In [40]: sys.float_info
Out[40]: sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)
```

**Comment by ethanwhite** [comment]
@davharris

**Comment by rueuntal** [comment]
```
sys.float_info
sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)
```
These look consistent.

**Comment by ethanwhite** [comment]
OK, let me set up a virtual environment w/0.15.1 and see if that makes any difference.

**Comment by ethanwhite** [comment]
I couldn't manage to get 0.15.1 up and running successfully... With 0.17.0 the first one works:

```
In [16]: ab = np.array([6, 2, 1, 1, 1, 4, 1, 1, 2, 1, 1])

In [17]: n, p = md.nbinom_lower_trunc_solver(ab)

In [18]: n, p
Out[18]: (0.00043828687940039454, 0.3078466861808869)

In [19]: md.nbinom_lower_trunc_ll(ab, n, p)
Out[19]: -14.094842869784218
```

but the 2nd still doesn't

```
In [25]: ab2 = np.array([  15,   12,    7,  295,    5,   66,   14,   48,   14,   71,   29,
         25,    1,  126,  102,    8,   69,   89,   35,    6,  111,    3,
         13,    5,   21,   22,  145,   13,  115,   13,    7,  818,    1,
        157,   66,    2,   21,   57,   70,   19,   40,   25,  414,  263,
        614,  124,  355,  130,  250,    1, 1364,   74,    4,   40,   92,
         11,   75,   13,    7,  290,   32])

In [26]: n, p = md.nbinom_lower_trunc_solver(ab2)

In [27]: n, p
Out[27]: (5.4114869127404129e-16, 0.90875366962832926)

In [28]: md.nbinom_lower_trunc_ll(ab2, n, p)
Out[28]: inf
```

So, yes, it's probably scipy version differences. Sigh.

**Comment by rueuntal** [comment]
That's too bad. Having one site removed (out of how many thousands?) is probably no big ideal, but I'm slightly more concerned about the difference leaking into likelihood estimates, and at the moment we don't know if there's one version that's more accurate than the others. 

Would it be worth it to dig deeper into this, or would it be an overkill?

**Comment by davharris** [comment]
Do the estimates differ across versions for the finite cases?

Dave

**Comment by rueuntal** [comment]
Can we randomly pick a bunch of sites and run them through the three different versions? I can do the test in scipy 0.15.1.

**Comment by ethanwhite** [comment]
I'll do a full comparison today. I have the `0.17.0` results from yesterday. I'm currently running `0.16.1` for CBC and FIA, since those are the two toughest datasets. I'll post a full comparison for those two datasets once that run is done.

**Comment by rueuntal** [comment]
Sounds great. Thanks, Ethan!

**Comment by ethanwhite** [comment]
OK, here it is:
https://gist.github.com/ethanwhite/30640ba9421ab4c18d41

Short version, most cases don't differ. Some do and potentially by a fair bit.

**Comment by rueuntal** [comment]
Thanks. Based on the head entries it seems that negbin is the only one having this issue. Is that correct?

What's your take on this?

**Comment by davharris** [comment]
For what it’s worth, I don’t see anything in the 0.17 release notes that would account for the difference. https://scipy.github.io/devdocs/release.0.17.0.html <https://scipy.github.io/devdocs/release.0.17.0.html>

**Comment by ethanwhite** [comment]
I just updated the gist to show the range for each model. Log-series and zipf have no variation at all, pln is so tiny it doesn't matter, so yes, just the negbin.

**Comment by davharris** [comment]
Looks like the optimizer did change a tiny bit between versions (e.g. https://github.com/fabian-paul/scipy/commit/a5ba086dcf7e9a8286796c2005d54818d0b5a9f8 <https://github.com/fabian-paul/scipy/commit/a5ba086dcf7e9a8286796c2005d54818d0b5a9f8>), but not enough to account for this, I don’t think.  Couldn’t find any substantive changes in nbinom.

**Comment by ethanwhite** [comment]
I'm currently rerunning w/0.17 again just to double check.

**Comment by ethanwhite** [comment]
This morning I tried to do a more controlled test of things using CBC. I ran everything on one computer, out of different directories, using different `conda` environments. I also ran `scipy` `0.17.0` twice in case what we were seeing were non-convergence issues.

Assuming I did everything right (and the histories on the 4 shells involved look good) then I'm now getting exactly the same results for `0.16.1` and `0.17.0` and the results for `0.17.0` are consistent across two runs. All three runs contain the `inf` record.

So, I'm not sure what happened. The previous comparison used two different computers (but same OS). That seems unlikely to be cause. There was also some definite potential for user error in how I did it. The most likely thing I could have done wrong would have been to run `0.15.1` accidentally due to not being careful about which environment I was running in (and it turns out I have `0.15.1` installed in the base environment on my desktop). I'm running this now to check. I could also have been running the wrong version of `macroecotools`, but that seems unlikely since the state of the repo was synced across my machines.

**Comment by davharris** [comment]
I think my R code that gets run in `check.py` is reliable.  Could we see what happens when that code is run on the results from 0.16 or 0.17?  

If macroecotools + current scipy agree with the R code, then I think we're good for the negative binomial.  If not, it's possible there are problems on the R end of things and I can look into them.

**Comment by ethanwhite** [comment]
Yep. I just started a full run on 0.17. When it finishes in a couple of hours I'll move the the output files into `sad-data` and the check.

**Comment by ethanwhite** [comment]
```
(sadscipy17)ethan@stice:~/Dropbox/Research/SADComparison (refactor)$ python check.py 
checking outputs:

Error: all.equal(my_ll, ll$likelihood_negbin, check.attributes = FALSE,  .... is not TRUE
Execution halted
Error in file(file, "rt") : cannot open the connection
Calls: read.csv -> read.table -> file
In addition: Warning message:
In file(file, "rt") :
  cannot open file 'sad-data/cbc_spab.csv': No such file or directory
Execution halted
Error: all.equal(my_ll, ll$likelihood_negbin, check.attributes = FALSE,  .... is not TRUE
Execution halted
Error: all.equal(my_ll, ll$likelihood_negbin, check.attributes = FALSE,  .... is not TRUE
Execution halted
Error: all.equal(my_ll, ll$likelihood_negbin, check.attributes = FALSE,  .... is not TRUE
Execution halted
Error in file(file, "rt") : cannot open the connection
Calls: read.csv -> read.table -> file
In addition: Warning message:
In file(file, "rt") :
  cannot open file 'sad-data/naba_spab.csv': No such file or directory
Execution halted

checking likelihood consistency:

  no problems found with likelihood consistency from  bbs 
Error: all.equal(my_relative, relative, check.attributes = FALSE) are not all TRUE
Execution halted
  no problems found with likelihood consistency from  fia 
  no problems found with likelihood consistency from  gentry 
  no problems found with likelihood consistency from  mcdb 
  no problems found with likelihood consistency from  naba 
```

**Comment by ethanwhite** [comment]
Updated the output after dealing with missing `_spab` file(s).

**Comment by davharris** [comment]
Is this solved? I don't have any NAs in the results from #191.

**Comment by ethanwhite** [comment]
@rueuntal can you run this as well?

```
python check.py
```

If the output is clean for you then go ahead and close this. If you get errors post the output here.

Thanks.

**Comment by rueuntal** [comment]
Using files from the Dropbox folder here's what I got:
```
$ python check.py
checking outputs:

  no problems found with output from  bbs
Error in Ops.factor(left, right) : level sets of factors are different
Calls: stopifnot -> Ops.data.frame -> eval -> eval -> Ops.factor
Execution halted
Error: all.equal(my_ll, ll$likelihood_negbin, check.attributes = FALSE,  .... is not TRUE
Execution halted
Error: all.equal(my_ll, ll$likelihood_negbin, check.attributes = FALSE,  .... is not TRUE
Execution halted
  no problems found with output from  mcdb
Error in Ops.factor(left, right) : level sets of factors are different
Calls: stopifnot -> Ops.data.frame -> eval -> eval -> Ops.factor
Execution halted

checking likelihood consistency:

  no problems found with likelihood consistency from  bbs
  no problems found with likelihood consistency from  cbc
Error: all.equal(my_relative, relative, check.attributes = FALSE) are not all TRUE
Execution halted
Error: all.equal(my_relative, relative, check.attributes = FALSE) are not all TRUE
Execution halted
  no problems found with likelihood consistency from  mcdb
Error: all.equal(my_relative, relative, check.attributes = FALSE) are not all TRUE
Execution halted
```

**Comment by davharris** [comment]
Thanks Xiao.

Not sure what’s up with the factors, but I suspect I just need to set stringsAsFactors=FALSE.

The negbin errors can be ignored (and those tests should be turned off) because we’re doing them in R now.

The issues with relative likelihood consistency are probably inherited from the negative binomial problems.

I should be able to fix all of these tomorrow.

**Comment by davharris** [comment]
Looks like essentially all of the checks I wrote are no longer relevant.  The negative binomial, AICc, and model weights are being done in postprocessing now.

**Comment by ethanwhite** [comment]
Great. If you're happy go ahead and close this issue.
